plutus-core/cost-model/data/builtinCostModelB.json

```diff
@@ -0,0 +1,919 @@
+            "type": "subtracted_sizes"
+        }
+    },
+    "encodeUtf8": {
+        "cpu": {
+            "arguments": {
+                "intercept": 1000,
+                "slope": 42921
+            },
+            "type": "linear_in_x"
+        },
+        "memory": {
+            "arguments": {
+                "intercept": 4,
+                "slope": 2
+            },
+            "type": "linear_in_x"
+        }
+    },
+    "equalsByteString": {
+        "cpu": {
+            "arguments": {
+                "constant": 24548,
+                "intercept": 29498,
+                "slope": 38
+            },
+            "type": "linear_on_diagonal"
+        },
+        "memory": {
+            "arguments": 1,
+            "type": "constant_cost"
+        }
+    },
+    "equalsData": {
+        "cpu": {
+            "arguments": {
+                "intercept": 898148,
+                "slope": 27279
+            },
+            "type": "min_size"
+        },
+        "memory": {
+            "arguments": 1,
+            "type": "constant_cost"
+        }
+    },
+    "equalsInteger": {
+        "cpu": {
+            "arguments": {
+                "intercept": 51775,
+                "slope": 558
+            },
+            "type": "min_size"
+        },
+        "memory": {
+            "arguments": 1,
+            "type": "constant_cost"
+        }
+    },
+    "equalsString": {
+        "cpu": {
+            "arguments": {
+                "constant": 39184,
+                "intercept": 1000,
+                "slope": 60594
+            },
+            "type": "linear_on_diagonal"
+        },
+        "memory": {
+            "arguments": 1,
+            "type": "constant_cost"
+        }
+    },
+    "fstPair": {
+        "cpu": {
+            "arguments": 141895,
+            "type": "constant_cost"
+        },
+        "memory": {
+            "arguments": 32,
+            "type": "constant_cost"
+        }
+    },
+    "headList": {
+        "cpu": {
+            "arguments": 83150,
+            "type": "constant_cost"
+        },
+        "memory": {
+            "arguments": 32,
+            "type": "constant_cost"
+        }
+    },
+    "iData": {
+        "cpu": {
+            "arguments": 15299,
+            "type": "constant_cost"
+        },
+        "memory": {
+            "arguments": 32,
+            "type": "constant_cost"
+        }
+    },
+    "ifThenElse": {
+        "cpu": {
+            "arguments": 76049,
+            "type": "constant_cost"
+        },
+        "memory": {
+            "arguments": 1,
+            "type": "constant_cost"
+        }
+    },
+    "indexByteString": {
+        "cpu": {
+            "arguments": 13169,
+            "type": "constant_cost"
+        },
+        "memory": {
+            "arguments": 4,
+            "type": "constant_cost"
+        }
+    },
+    "integerToByteString": {
+        "cpu": {
+            "arguments": {
+                "c0": 9223372036854775807,
```

Comment:
As discussed, `integerToByteString` and `byteStringToInteger` should have normal parameters.

Reply:
I'm not sure about that because this model will apply from PV9 but those builtins won't be available until PV10.

Reply:
> I'm not sure about that because this model will apply from PV9 but those builtins won't be available until PV10.

To expand on that, the actual numbers in the file don't matter because the node will overwrite by the protocol parameters (which in fact we'll probably provide, based on the contents of this file).  In PV9 these builtins won't be available but in PV10 we will have real values for these parameters.  As I mentioned earlier we might strictly want four cost models, with one (based on this one) for PV9 and an updated version for PV10 (although it's probably safe to have sensible numbers here because they won't be used anyway).  We'll maybe be adding more builtins for PV10, so we'll have to update this for that anyway.

Reply:
> which in fact we'll probably provide, based on the contents of this file

That would be a problem. When we provide the parameters, all parameters should be normal (non-max). Otherwise, if we enable a builtin at PV10, then before the PV9 -> PV10 transition, its parameters must be updated from max to normal, which is just an unnecessary burden.

Also, note that in the Conway era, it may not even be possible to update a parameter exactly at the same time of a HF, due to new governance mechanisms. If so, we'll have to update the parameters before the HF. But if we do that, we may as well just provide the normal parameters in the first place!

Reply:
> That would be a problem. When we provide the parameters, all parameters should be normal (non-max).

I thought we wouldn't provide those parameters until the buitlins were actually available, so the PV9 parameters for PlutusV1/2 wouldn't contain any entries for `integerToByteString/byteStringToInteger` but the PV10 ones would.  Maybe you're right though: the PV9 protocol parameters could contain cost model parameters for those bultins but they'd never be used in PV9 because the builtins are disabled.  I suppose we might need to recalibrate the cost model in between PV9 and PV10 though, in which case we'd need to update the parameters anyway.

The procedure for updating the cost model parameters is a bit informal: we just come up with some numbers and occasionally we provide updated values to the node people for incorporation in the protocol parameters; in some PV updates nothing changes but in others the cost model parameters really do change.  Maybe we need some more formal process for doing this.

The reason that I put in the maximum numbers is that (at least with the way we currently decode the cost model from JSON) the JSON files _have_ to contain entries for all of the builtins, even for ones which aren't available during the relevant PV.  Those numbers are really supposed to serve as a `Nothing` value, and also have the advantage of making the builtins unusable for our off-chain tools and tests, where code doesn't always go trough the deserialisation process that checks for invalid builtins.  I think I said this somewhere else, but maybe a better way to do that would be to  serialise and deserialise with the appropriate PV/LL combination when we're using code that's read in from textual source or compiled from Haskell. That might be something to look into in #6088.

Reply:
Also, I really wish you could put comments in JSON files to explain what's going on.